rq/group.py

```diff
@@ -0,0 +1,126 @@
+            for job in job_ids:
+                p.exists(Job.key_for(job))
+            results = p.execute()
+
+        for i, key_exists in enumerate(results):
+            if not key_exists:
+                expired_job_ids.append(job_ids[i])
+        if expired_job_ids:
+            pipe.srem(self.key, *expired_job_ids)
+        if pipeline is None:
+            pipe.execute()
+
+    def enqueue_many(self, queue: Queue, job_datas: List['EnqueueData'], pipeline: Optional['Pipeline'] = None):
+        pipe = pipeline if pipeline else self.connection.pipeline()
+
+        jobs = queue.enqueue_many(job_datas, group_id=self.name, pipeline=pipe)
+
+        self._add_jobs(jobs, pipeline=pipe)
+
+        if pipeline is None:
+            pipe.execute()
+
+        return jobs
+
+    def get_jobs(self) -> list:
+        """Retrieve list of job IDs from the group key in Redis"""
+        self.cleanup()
+        job_ids = [as_text(job) for job in self.connection.smembers(self.key)]
+        return [job for job in Job.fetch_many(job_ids, self.connection) if job is not None]
+
+    def delete(self):
+        self.connection.delete(self.key)
+
+    def delete_job(self, job_id: str, pipeline: Optional['Pipeline'] = None):
+        pipe = pipeline if pipeline else self.connection.pipeline()
+        pipe.srem(self.key, job_id)
+        if pipeline is None:
+            pipe.execute()
+
+    @classmethod
+    def create(cls, connection: Redis, name: Optional[str] = None):
+        return cls(name=name, connection=connection)
+
+    @classmethod
+    def fetch(cls, name: str, connection: Redis):
+        """Fetch an existing group from Redis"""
+        group = cls(name=name, connection=connection)
+        if not connection.exists(Group.get_key(group.name)):
+            raise NoSuchGroupError
+        return group
+
+    @classmethod
+    def all(cls, connection: 'Redis') -> List['Group']:
+        "Returns an iterable of all Groupes."
+        group_keys = [as_text(key) for key in connection.smembers(cls.REDIS_GROUP_KEY)]
+        return [cls.fetch(key, connection=connection) for key in group_keys]
+
+    @classmethod
+    def get_key(cls, name: str) -> str:
+        """Return the Redis key of the set containing a group's jobs"""
+        return cls.REDIS_GROUP_NAME_PREFIX + name
+
+    @classmethod
+    def clean_registries(cls, connection: 'Redis'):
+        """Loop through groups and delete those that have been deleted.
+        If group still has jobs in its registry, delete those that have expired"""
+        groups = Group.all(connection=connection)
+        with connection.pipeline() as p:
+            if groups:
+                # Remove expired jobs from groups
+                for group in groups:
+                    group.cleanup(pipeline=p)
+                p.execute()
+                # Remove empty groups from group registry
+                for group in groups:
+                    p.exists(group.key)
```

Comment:
Should we move this logic into `group.cleanup()`?

Reply:
@selwin I think it makes more sense here since it's not related to any one group. Happy to move it into a separate classmethod if this one's getting to busy though.